HPCC-23433 Fix locking race condition with RTM_DELETE_ON_DISCONNECT

jakesmith opened the pull request:
When deleting the root node on a RTM_DELETE_ON_DISCONNECT, a
mutex was being released prematurely, allowing another thread
to delete the root of the connection at the same time.
This in turn caused a 'Transaction to orphaned server node'
exception to be thrown.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith committed:
HPCC-23433 Fix locking race condition with RTM_DELETE_ON_DISCONNECT

When deleting the root node on a RTM_DELETE_ON_DISCONNECT, a
mutex was being released prematurely, allowing another thread
to delete the root of the connection at the same time.
This in turn caused a 'Transaction to orphaned server node'
exception to be thrown.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-13300, label: hpcc-23433-rtm-lock-race starts now.
The reason of this test is: Did not build yet.
Commit ID: f4ed6b7692104a351452db0dedb4169eb3850ad3
Estimated completion time is ~0.5 hour(s)

jakesmith commented:
NB: a lot of the changes in this PR are cosmetic refactoring, resuling from the trivial CLCReadLockBlock, CLCWriteLockBlock classes being deleted in favour of their base class, so as to make it simpler to call the new changeToWrite method in the ReadWriteLock.

@mckellyln - please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: f4ed6b7692104a351452db0dedb4169eb3850ad3
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |863 |863 |0 |456 sec  (00:07:36) |
|test (thor) |802 |802 |0 |879 sec  (00:14:39) |
|test (roxie) |944 |944 |0 |580 sec  (00:09:40) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) |  741 sec (00:12:21) |  119 sec (00:01:59) |   21 sec (00:00:21) |   17 sec (00:00:17) | 2223 sec (00:37:03) |   18 sec (00:00:18) | 3153 sec (00:52:33) |

HPCCSmoketest commented:
Process of PR-13300, label: hpcc-23433-rtm-lock-race starts now.
The reason of this test is: New commit.
Commit ID: 90e4b77c65061084ca0094bce47d3b2d791a492c
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 90e4b77c65061084ca0094bce47d3b2d791a492c
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |18 sec  (00:00:18) |
|test (hthor) |863 |863 |0 |451 sec  (00:07:31) |
|test (thor) |802 |802 |0 |877 sec  (00:14:37) |
|test (roxie) |944 |944 |0 |578 sec  (00:09:38) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  733 sec (00:12:13) |  119 sec (00:01:59) |   22 sec (00:00:22) |   17 sec (00:00:17) | 2236 sec (00:37:16) |   19 sec (00:00:19) | 3154 sec (00:52:34) |

mckellyln commented on system/jlib/jmutex.hpp:
```diff
@@ -586,67 +586,88 @@ class jlib_decl Monitor: public Mutex
+        }
+        if (readLocks)
+            readLocks--;
+        return doLockWrite(timed, timeout); // returns false if timed out
```
Should we try the doLockWrite() and then if ok decrement readLocks ?

jakesmith replied:
yes, but in fact that introduces a whole in itself. It should not release/reduce the readLocskc count here 1st.
Unfortunately, that makes the implementation trickier in that there may be something in a new waiting state.
I'll revisit.

mckellyln commented on system/jlib/jmutex.hpp:
```diff
@@ -586,67 +586,88 @@ class jlib_decl Monitor: public Mutex
+            cs.leave();
+            if (timed)
+            {
+                if (!readSem.wait(timeout)) {
```
Might as well move these parens as well

jakesmith replied:
yep agree, will change formatting here, and below.

mckellyln commented:
@jakesmith one question and one comment.

HPCCSmoketest commented:
Process of PR-13300, label: hpcc-23433-rtm-lock-race starts now.
The reason of this test is: New commit.
Commit ID: de185eae3408b308d77d2e363e4411af1a821dc7
Estimated completion time is ~0.9 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: de185eae3408b308d77d2e363e4411af1a821dc7
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |199 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |18 sec  (00:00:18) |
|test (hthor) |868 |868 |0 |453 sec  (00:07:33) |
|test (thor) |807 |807 |0 |883 sec  (00:14:43) |
|test (roxie) |949 |949 |0 |606 sec  (00:10:06) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  743 sec (00:12:23) |  119 sec (00:01:59) |   22 sec (00:00:22) |   17 sec (00:00:17) | 2440 sec (00:40:40) |   19 sec (00:00:19) | 3368 sec (00:56:08) |

jakesmith commented:
@ghalliday @mckellyln - more complicated than I had initially anticipated (in ReadWriteLock)

Please review.

ghalliday commented on system/jlib/jmutex.hpp:
```diff
@@ -661,35 +710,74 @@ class jlib_decl ReadWriteLock
+    void unlock()
+    {
+        cs.enter(); 
+        if (readLocks)
```
Could do with some comments.  E.g. if (readLocks is non zero then a read lock must be being released)

jakesmith replied:
will add some comments.

ghalliday commented on system/jlib/jmutex.hpp:
```diff
@@ -661,35 +710,74 @@ class jlib_decl ReadWriteLock
+    bool changeToRead()
+    {
+        cs.enter();
+        if (!writeLocks)
```
minor - this could be checked outside the critical section.  Should it be an internal error rather than returning false?

jakesmith replied:
I think a boolean result indicating not locked failure, similar to the failure on timed lock read/write calls.

>this could be checked outside the critical section

I'm not sure this edge case makes it worth moving the test, and clearer semantics if inside crit ?

ghalliday replied:
I don't really mind if it is inside or outside.
If the function should never be called in that situation then it should be an internal error.  If it is expected then there should be a comment e.g.
//If the lock is already a read lock then this function will return false.

Similarly for changeToWrite - if the lock is already a write lock then it will return false.
Is that the correct response though?  If there is a timeout you can't distinguish between timedout and couldn't get the lock, and already had the write lock.

I don't have a strong view on which semantics you choose, but it should have a comment documenting what they are.

jakesmith replied:
>If there is a timeout you can't distinguish between timedout and couldn't get the lock, and already had the write lock.

ok yes I agree, on reflection I think these unexpected use cases should throw an exception instead. Will change.

ghalliday commented on system/jlib/jmutex.hpp:
```diff
@@ -586,71 +586,120 @@ class jlib_decl Monitor: public Mutex
+    bool changeToWrite(bool timed, unsigned timeout)
+    {
+        cs.enter();
+        if (writeLocks || (0 == readLocks))
```
more: should this  throw an internal error instead of returning false?  ((readLocks==0) could be checked before the cs.

ghalliday replied:
or is the intention that it is ok to call this if it already has a write lock.  In which case the comment on the return is misleading.

jakesmith replied:
it's supposed to fail if already has a write lock, but a soft failure (return false) as opposed to an exception - in keeping with pre-exisitng timed lock read / write calls that return false on failure to get lock.

jakesmith replied:
>((readLocks==0) could be checked before the cs.

should it though ? it's an edge case, not generally expected, and I think code clearer when checked inside the crit.

ghalliday replied:
The comment implies it is not legal to call it when not read locked.  If that is kind-of allowed the comment should be changed.  E.g.
//if called when not readlocked then fail immediately.

ghalliday reviewed (commented):
@jakesmith I have looked at it and the change looks good.  (Although the code in jlib could do with more comments, and I'm not convinced it is good for it to all be inline).
@mckellyln I think it is worth you checking carefully as well.

jakesmith commented:
@ghalliday - thanks, added some replies and questions back re. whether some of these conditions really should be moved before locking the crit.

jakesmith commented:
.. and @mckellyln - please also review.

HPCCSmoketest commented:
Process of PR-13300, label: hpcc-23433-rtm-lock-race starts now.
The reason of this test is: New commit.
Commit ID: 3374a1ef64d46267676c07f8a5b8d93fc7c7ea95
Estimated completion time is ~0.9 hour(s)

jakesmith commented:
@mckellyln @ghalliday - changed invalid use cases to exceptions and added comments.

Have also moved some inline methods out of header into cpp, partly because use of jexcept inside header was causing circular header dependency issues.

mckellyln commented:
@jakesmith ok, reviewiing still

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 3374a1ef64d46267676c07f8a5b8d93fc7c7ea95
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |199 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |868 |868 |0 |458 sec  (00:07:38) |
|test (thor) |807 |807 |0 |880 sec  (00:14:40) |
|test (roxie) |949 |949 |0 |578 sec  (00:09:38) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  748 sec (00:12:28) |  119 sec (00:01:59) |   22 sec (00:00:22) |   17 sec (00:00:17) | 2444 sec (00:40:44) |   18 sec (00:00:18) | 3376 sec (00:56:16) |

mckellyln commented on system/jlib/jmutex.cpp:
```diff
@@ -321,7 +321,191 @@ void Monitor::notifyAll()
+    {
+        writeLocks--;
+        dbgassertex(writeLocks == 0);
+        if (readWaiting)
```
So we are ok with readers always acquiring before any writers ?
Perhaps the check for writers should come first ?  Or we could rotate the order of this check on each unlock ?

jakesmith replied:
not sure, but that is not new semantics, ReadWriteLock has always given priority to pending blocked readers over blocked writers.

Worth considering, but as a separate task/discussion/JIRA.

jakesmith commented:
>@jakesmith ok, reviewiing still

@mckellyln - re-tagging

mckellyln reviewed (approved):
I am ok with this, but I do think writers should get priority over readers.

mckellyln commented:
@jakesmith ok for merge.

jakesmith commented:
>I am ok with this, but I do think writers should get priority over readers.

as I say, worth considering, but wouldn't want change the pre-existing semantics in this PR.
Please open new JIRA to look at / further discussion.

jakesmith commented:
@ghalliday - will squash ready for merge.

jakesmith commented:
@ghalliday - please merge.

HPCCSmoketest commented:
Process of PR-13300, label: hpcc-23433-rtm-lock-race starts now.
The reason of this test is: New commit.
Commit ID: 430526974c6a577ae29abc3d51854e3a219e23b9
Estimated completion time is ~0.9 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 430526974c6a577ae29abc3d51854e3a219e23b9
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |199 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |871 |871 |0 |457 sec  (00:07:37) |
|test (thor) |810 |810 |0 |870 sec  (00:14:30) |
|test (roxie) |951 |951 |0 |599 sec  (00:09:59) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  745 sec (00:12:25) |  115 sec (00:01:55) |   25 sec (00:00:25) |   17 sec (00:00:17) | 2364 sec (00:39:24) |   19 sec (00:00:19) | 3302 sec (00:55:02) |

ghalliday commented on system/jlib/jmutex.cpp:
```diff
@@ -321,7 +321,191 @@ void Monitor::notifyAll()
+bool ReadWriteLock::changeToRead()
+{
+    cs.enter();
+    if (!writeLocks) // NB: If the lock is already a read lock then this function will return false.
```
I'm not sure the comment is correct

jakesmith replied:
no, should have been removed.

ghalliday reviewed (commented):
@jakesmith one minor comment on a comment

jakesmith commented:
@ghalliday - removed rogue comment.
